core/arch/arm/plat-hikey/spi_test.c

```diff
@@ -28,73 +28,279 @@
 
-	DMSG("gpio6_base: 0x%" PRIxVA "\n", gpio6_base);
 	DMSG("spi_base: 0x%" PRIxVA "\n", spi_base);
+	DMSG("configure SPI\n");
+	platform_pl022_data.base = spi_base;
+	platform_pl022_data.clk_hz = SPI_CLK_HZ;
+	platform_pl022_data.speed_hz = SPI_10_KHZ;
+	platform_pl022_data.mode = SPI_MODE0;
+	platform_pl022_data.data_size_bits = 8;
+	platform_pl022_data.loopback = true;
+
+	pl022_init(&platform_pl022_data);
+	platform_pl022_data.chip.ops->configure(&platform_pl022_data.chip);
+	platform_pl022_data.chip.ops->start(&platform_pl022_data.chip);
+
+	for (j = 0; j < 20; j++) {
+		DMSG("SPI test loop: %zu\n", j);
+		res = platform_pl022_data.chip.ops->txrx8
+			(&platform_pl022_data.chip, tx, rx, len);
+
+		if (res) {
+			DMSG("SPI transceive error %d\n", res);
+			break;
+		}
 
+		for (i = 0; i < len; i++)
+			DMSG("rx[%zu] = 0x%x\n", i, rx[i]);
+
+		tee_time_busy_wait(20);
+	}
+
+	platform_pl022_data.chip.ops->end(&platform_pl022_data.chip);
+}
+
+static void spi_test_with_builtin_cs_control(void)
+{
+	struct pl061_data platform_pl061_data;
+	struct pl022_data platform_pl022_data;
+	vaddr_t gpio6_base = nsec_periph_base(GPIO6_BASE);
+	vaddr_t spi_base = nsec_periph_base(SPI_BASE);
+	uint8_t tx[3] = {0x01, 0x80, 0x00};
+	uint8_t rx[3] = {0};
+	size_t i, j, len = 3;
+	enum spi_result res;
+
+	spi_set_cs_mux(PINMUX_GPIO);
+
+	DMSG("gpio6_base: 0x%" PRIxVA "\n", gpio6_base);
 	DMSG("configure GPIO\n");
 	pl061_init(&platform_pl061_data);
 	pl061_register(gpio6_base, 6);
-
 	DMSG("enable software mode control for chip select\n");
 	pl061_set_mode_control(GPIO6_2, PL061_MC_SW);
 
-	DMSG("mask/disable interrupt for chip select\n");
-	platform_pl061_data.chip.ops->set_interrupt(GPIO6_2,
-						GPIO_INTERRUPT_DISABLE);
+	platform_pl022_data.cs_data.gpio_data.chip =
+		&platform_pl061_data.chip;
+	platform_pl022_data.cs_data.gpio_data.pin_num = GPIO6_2;
+	platform_pl022_data.b_cs_callback = false;
 
+	DMSG("spi_base: 0x%" PRIxVA "\n", spi_base);
 	DMSG("configure SPI\n");
-	platform_pl022_data.gpio = &platform_pl061_data.chip;
 	platform_pl022_data.base = spi_base;
-	platform_pl022_data.cs_gpio_base = gpio6_base;
 	platform_pl022_data.clk_hz = SPI_CLK_HZ;
-	platform_pl022_data.speed_hz = SPI_500_KHZ;
-	platform_pl022_data.cs_gpio_pin = GPIO6_2;
+	platform_pl022_data.speed_hz = SPI_10_KHZ;
 	platform_pl022_data.mode = SPI_MODE0;
 	platform_pl022_data.data_size_bits = 8;
 	platform_pl022_data.loopback = true;
 
-	pl022_configure(&platform_pl022_data);
-	pl022_start(&platform_pl022_data);
+	pl022_init(&platform_pl022_data);
+	platform_pl022_data.chip.ops->configure(&platform_pl022_data.chip);
+	platform_pl022_data.chip.ops->start(&platform_pl022_data.chip);
 
 	for (j = 0; j < 20; j++) {
 		DMSG("SPI test loop: %zu\n", j);
-		platform_pl022_data.chip.ops->txrx8(&platform_pl022_data.chip,
-						tx, rx, len, &num_rxpkts);
-		for (i = 0; i < num_rxpkts; i++)
+		res = platform_pl022_data.chip.ops->txrx8
+			(&platform_pl022_data.chip, tx, rx, len);
```

Comment:
I would prefer the parenthesis to be right next to the handler (@jerome: your opinion)
```
		res = platform_pl022_data.chip.ops->txrx8(
					&platform_pl022_data.chip, tx, rx, len);
```
maybe you could find a way to prevent such line splits.